File: e086(computer_scoring_2)/Java/ComputerScoring.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.HashMap;


public class ComputerScoring {
    public static void main(String[] args) throws IOException {
        try (BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
             BufferedWriter out = new BufferedWriter(new OutputStreamWriter(System.out))) {
            String[] numLine = in.readLine().split(" ");
            int numOfSub = Integer.parseInt(numLine[0]);
            int numOfStd = Integer.parseInt(numLine[1]);

            /*
             * student numbers are always in 3 digits, while subject numbers are not
             * subNum(int) to subNum(String) mapping, index is the int form (0 not used)
             */
            String[] strSubs = new String[numOfSub + 1];
            /* stores subNum : (subName, correctAns), index is the subNum, 0 not used */
            SubAndAns[] subs = new SubAndAns[numOfSub + 1];
            for (int subNum = 1; subNum <= numOfSub; ++subNum) {
                String[] dataLine = in.readLine().split(" ");
                strSubs[subNum] = dataLine[0];
                subs[subNum] = new SubAndAns(dataLine[1], dataLine[2].toCharArray());
            }

            /*  stores stdNum : stdName, index is the stdNum, 0 not used */
            String[] stds = new String[numOfStd + 1];
            for (int stdNum = 1; stdNum <= numOfStd; ++stdNum) {
                String[] dataLine = in.readLine().split(" ");
                stds[stdNum] = dataLine[1];
            }

            /*
             * stores stdNum + subNum : score,
             * trick at using String concatenation as the key
             */
            HashMap<String, Integer> scores = new HashMap<>();
            for (int i = 0; i < numOfStd * numOfSub; ++i) {
                String[] dataLine = in.readLine().split(" ");
                char[] correctAns = subs[Integer.parseInt(dataLine[1])].correctAns;
                char[] ans = dataLine[2].toCharArray();
                scores.put(
                    dataLine[0] + dataLine[1],
                    calculateScore(correctAns, ans)
                );
            }

            /* output formatting: stdNum,stdName,subNum,subName,score */
            for (int stdNum = 1; stdNum <= numOfStd; ++stdNum) {
                String strStdNum = String.format("%03d", stdNum);
                for (int subNum = 1; subNum <= numOfSub; ++subNum) {
                    /* student part */
                    out.write(
                        String.format("%s,%s,", strStdNum, stds[stdNum])
                    );
                    /* subject and score part*/
                    out.write(
                        String.format("%s,%s,%d\n", strSubs[subNum], subs[subNum].subName,
                                                    scores.get(strStdNum + strSubs[subNum]))
                    );

                }
            }
            out.flush();
        }  /* try-with-resources */
    }  /* main */


    /*
     * To save time and memory (not using too many Maps),
     * customized class is used to let a Map stores 3 data.
     */
    static class SubAndAns {
        public final String subName;
        public final char[] correctAns;

        public SubAndAns(final String subName, final char[] correctAns) {
            this.subName = subName;
            this.correctAns = correctAns;
        }
    }


    static int calculateScore(final char[] correctAnswer, final char[] answer) {
        /* counter */
        int correct = 0;
        /* too much not count, not enough no penalty */
        int valids = (answer.length > correctAnswer.length
        ? correctAnswer.length
        : answer.length);
        for (int i = 0; i < valids; ++i) {
            if (answer[i] == correctAnswer[i]) {
                ++correct;
            }
        }
        return (correct * 100 / correctAnswer.length);
    }

}  /* ComputerScoring */

File: e086(computer_scoring_2)/Python/computer_scoring.py
from sys import stdin, stdout
from typing import Dict, List, Tuple


def calculate_score(correct_ans: str, ans: str) -> int:
    correct: int = 0
    for i in range(len(ans) if len(correct_ans) > len(ans) else len(correct_ans)):
        if (ans[i] == correct_ans[i]):
            correct += 1
    return correct * 100 // len(correct_ans)


for line in stdin:
    if not line.rstrip('\n'):
        break
    num_of_subs, num_of_stds = map(int, line.split())
    # sub_num : [sub_name, sub_ans]
    subs: Dict[int, Tuple[str, str]] = {}
    # sub_num(int) to sub_num(str) mapping, index is the int form (0 not used)
    sub_int_str: List[str] = [""] * (num_of_subs + 1)
    for s in range(1, num_of_subs + 1):
        sub_num, sub_name, ans = stdin.readline().split()
        sub_int_str[s] = sub_num
        subs[s] = (sub_name, ans)

    # std_num : std_name
    roster: Dict[int, str] = {}
    # std_num : (sub_num, score)
    std_score: Dict[int, List[Tuple[int, int]]] = {}
    for s in range(1, num_of_stds + 1):
        std_num, std_name = stdin.readline().split()
        roster[s] = std_name
        std_score[s] = []

    for _ in range(num_of_subs * num_of_stds):
        std_num, sub_num, ans = stdin.readline().split()
        correct_ans: str = subs[int(sub_num)][1]
        std_score[int(std_num)].append(
            (int(sub_num), calculate_score(correct_ans, ans))
        )

    for std in range(1, num_of_stds + 1):
        std_score[std].sort(key=lambda s: s[0])
        for sub in range(1, num_of_subs + 1):
            stdout.write(f'{std:03d},{roster[std]},{sub_int_str[sub]},{subs[sub][0]},{std_score[std][sub - 1][1]}\n')
